### ISSUE TITLE
token uniqueness (branch 2.0.x)

### ISSUE THREAD

**Issue by boaks** [eclipse/californium#500]
We discussed this topic already in other issues, with long histories. I hope my sum-up therefore contains the important information. If not, feel free to add that.

Currently I see four variants:
1. Create token unique in the scope of an endpoint
1. Create token unique in the scope of the system
1. Create token random (and assume uniqueness in the scope of an endpoint)
1. Create token random (and assume uniqueness in the scope of the system)

The random variant will not require to do the housekeeping, the none-random will exclude extremely rare cases of token clashes (or allow to use smaller tokens, if wanted). 

Currently a mixture is implemented: create with 1. but partially used as 2 (ObservationStore), so in facto it's 4 :-).

To limit the changes, I would go for a clear 2, and take the house-keeping, mainly because I think, right now the housekeeping must be done anyway in the observe- and exchange-stores. 
Using approach 2 alone (system unique tokens) will not be enough, see issue #173. Therefore, an additional verification would be required for the request/response matching. Though the EndpointContextMatcher does this, I belief, this approach of handling tokens will be the one with the fewest changes. So I would go to prepare a PR as base for a further discussion.

Using approach 3 seems for me currently to result in some more changes, but feel free to provide an alternative PR with that.

**Comment by sbernard31** [comment]
The 2. variant force us to go for a reserve/release pattern. Housekeeping is one of the "con" but not the only one. In cluster mode this will force to go for a shared tokenprovider.

The issue with 3. is that making token unique for a given endpoint bring us to define what is an endpoint identifier, we partially address it with endpointContext.

Using approach 3. solves this is issue in a more stateless way.
So I still feel that 3. is the better one. I agree this will make probably more changes. But I also know that housekeeping and race condition hunting is timeconsuming too.

About proposing an alternative PR for 3, I could go for this only if there is chances that was accepted. My last experience on this topic discourages me a bit. Lot of time to code, lot of time to argue for nothing :'(

**Comment by boaks** [comment]
> The 2. variant force us to go for a reserve/release pattern.

FMPOV it allows us to `keep` not force us to `go`. 
To get rid of the housekeeping, it must be removed from the observe- and exchange-stores also. Or what are your plans for that stores?

I also thought about making variant 2 and 3 both possible with the interfaces and allow different implementations. 

If the `TokenProvider` and `KeyToken` uses a "byte[] identity" (and null or empty should be supported) and the `EndpointContextMatcher` is extended with "byte[] getIdentity(EndpointContext)", so that the `EndpointContextMatcher` implementation can decide on what the identity is based on.

The implementation of the `TokenProvider` should then be configurable, and you may chose one, which just generates random keys without housekeeping. The penalty would be, that `releaseToken` must still be called, even if the pure random implementation doesn't require it.  And the nasty will be, that we then could only create the "KeyToken" after the `onContextEstablished` :-). 
So I'm not sure, if such an implementation will pay off.

So, let me start with the simple approach, and I see, if I can prepare the combined approach next week as well.

**Comment by sbernard31** [comment]
It sounds a good plan.

Just to let you know, I faced users which are behind a NAT and which answer to request with a separate response. They ACK quickly then send the response long time after. Because of NAT the IP changed and with the current implementation response was ignore because KeyToken use "address/port" instead of "endpoint identifier".

I suppose your modifications should solves this right ? Even the simple one as you should remove address from KeyToken ?

**Comment by boaks** [comment]
> Just to let you know, I faced users which are behind a NAT and which answer to request with a separate response. They ACK quickly then send the response long time after. Because of NAT the IP changed and with the current implementation response was ignore because KeyToken use "address/port" instead of "endpoint identifier".

Using coap, this will be somehow a risc, because then only the token could be used and I don't see the second artefact to prove the matching.

Using coaps, the response may be just ignored because of a invalid MAC. But maybe, if PR #498 is used, it may work.

**Comment by sbernard31** [comment]
>  Using coap, this will be somehow a risc
by definition using coap means risk :)

**Comment by boaks** [comment]
> by definition using coap means risk :)

Using the token only would make the risc larger.

**Comment by boaks** [comment]
After thinking about the interfaces to support both, variant 2 and 3, it seems for me to be only useful, if a principal or address is provided (and using the address means, address changes are not supported). 
Using the session would be challenging, because the token is generated before the session is created. That would require to implement a "connect" (to establish the session and then create the KeyToken using the session information). On sending, there may be the possibility to dynamically select the additional information for KeyToken (e.g. use the session, but if, not available, the address), but on receiving the responses, this doesn't work without additional information. A solution may be, to use the first 2 bit to encode the scope (00 = system, 01 = principal, 10 = session, and 11 = address). But this must then be documented for application layer tokens (these first two bits are overwritten).     

So I would start with the simple ones, address and principal ...

**Comment by sbernard31** [comment]
When you are using 3, you don't need to have context to generate token because this is done randomly.
When you are using 2, you don't need to have context to generate token because token is unique overall the system.
So you can generate token before to get context.

But for 3. you need context to be able to store is in MessageExchangeStore or ObservationStore.

So a solution could be : 
1) generate token (no context)
2) send data through connector
3) on contextEstablished, add observation or exchange to stores.

I think this is compatible with this proposed modification : https://github.com/eclipse/californium/issues/487

**Comment by boaks** [comment]
Terms:
`Token`: 0-8 bytes used by a CoAP Request/Response
`KeyToken`: pair of `Token` and "identity" (bytes derived from `EndpointContext` using a `EndpointContextMatcher`)

Though the Exchange requires a `KeyToken` to be stored in the `MessageExchangeStore`, variant 3 requires a `EndpointContext` where a "identity" could be derived from. and therefore it get's complex :-).

Currently I'm working on my proposal for such API's, which support both implementation (variant 2 and 3). But it requires more changes than I assumed. I'm not sure, if I can finish the proposal today. But tomorrow should be possible.

**Comment by sbernard31** [comment]
I'm not sure you get my point. So I try to re-explain it differently.

`Token` generation does not need `EndpointContext` for 2. and 3. (**even for sessionID**)
This `Token` can be used to send rawdata throught connector.
We need `EndpointContext` to create `KeyToken` to add exchange or observation to store, and this can be done on `contextEstablished` event. (#487)

(I mean no need to implement "connect")

**Comment by boaks** [comment]
Which would require even more changes in the MessageExchangeStore :-).
Let's wait for tomorrow and then we may see, what will be required to changed that,

**Comment by boaks** [comment]
The variant, which stores exchanges and observes on "contextEstablished" causes really a flood of changes. And a there are also some critical changes, where I'm not clear, if the changed implementation is right.

Therefore, @sbernard31, can you please tell us, what the benefit of no house-keeping on the tokens should be, if the exchanges and observes are still require such a house-keeping?
I'm currently more afraid from new introduced issues by those changes than by the current ones.

Tomorrow I have too much other work, and on friday I will go earlier. So I'm not sure, if I can finish that work on that variant before Christmas.

**Comment by sbernard31** [comment]
> Therefore, @sbernard31, can you please tell us, what the benefit of no house-keeping on the tokens should be, if the exchanges and observes are still require such a house-keeping?

No housing-keeping is just one of the benefits (probably the smaller as like you said we need to already do that for exchange and observation)

The other benefits are :
- stateless code is more thread friendly (hunting race condition is a hard job)
- stateless is also cluster friendly. (no need to share state and synchronize access between instances)

> I'm currently more afraid from new introduced issues by those changes than by the current ones.

#506 is good enough to support LWM2M use case in a stateless way. (If I didn't miss something)
So if you think going deeper, (I mean supporting session ID as identifier) is too timeconsuming and too risky for poor benefits, we can keep this last step for later and see if users ask for him later ?

> Tomorrow I have too much other work, and on friday I will go earlier. So I'm not sure, if I can finish that work on that variant before Christmas.

No problem, I will not be there from 12/22/17 to 01/02/18 (included).

**Comment by boaks** [comment]
Currently one hotspot is the `CoapObserveRelation`:

```
	private void cancel() {
		endpoint.cancelObservation(...);
		setCanceled(true);
	}
``` 

this ends up in `BaseMatcher`:

```
	public void cancelObserve(...t) {
		for (Exchange exchange : exchangeStore.findByToken(token)) {
        ...
```

and so in the `InMemoryMessageExchangeStore`:

```
	public List<Exchange> findByToken(...) {
		List<Exchange> result = new ArrayList<>();
		if (token != null) {
			for (Entry<KeyToken, Exchange> entry : exchangesByToken.entrySet()) {
				if (entry.getValue().isOfLocalOrigin()) {
					Request request = entry.getValue().getRequest();
					if (request != null && token.equals(request.getToken())) {
						result.add(entry.getValue());
					}
				}
			}
		}
		return result;
	}
```

The first hot thing here is, this searches for a exchange of blockwise-notifies using the token of the first request, which is not longer mapped. But this has a strong impact on the performance, if we iterate over all exchanges. I guess we should improve this by a blockwise notify map filled in matchNotifyResponse.

The second hot thing here is, that this CoapObserveRelation would also require the EndpointContext to successful remove the Observation from the ObservationStore. And that seems for me to be also a candidate for race conditions, if `cancel()` and `contextEstablished` are processed in parallel.

**Comment by sbernard31** [comment]
> The first hot thing here is, this searches for a exchange of blockwise-notifies using the token of the first request, which is not longer mapped. But this has a strong impact on the performance, if we iterate over all exchanges. 

I agree this is really crappy.
This mainly aims to stop a blockwise exchange of a notification.
Ironically, this is a kind of "optimization" :D (I mean stopping unneeded block transfer)

> I guess we should improve this by a blockwise notify map filled in matchNotifyResponse.

Using a Map should works, but we can maybe attack this problem from another angle :
1. just left this optimization aside (this will change behavior a bit, but maybe this is acceptable)
2. changing paradigm : cancel observation is just removing from the store, up to blockwise layer to look at this to know the observation relation state. (this imply blockwiseLayer depends of ObservationStore  + more access to the store and this will not stop block2 retransmission of current request BUT this will  stop transfer on all cluster node)

> The second hot thing here is, that this CoapObserveRelation would also require the EndpointContext to successful remove the Observation from the ObservationStore. And that seems for me to be also a candidate for race conditions, if cancel() and contextEstablished are processed in parallel.

If we move contextEstablished on Message (#487), maybe this will be easier to handle this race condition?

**Comment by boaks** [comment]
After implementing 3 variants:
- PR #520 Plain Token (1000 changes)
- PR #519 Token and KeyToken, store in onContextEstablished (1800 changes)
- PR #518 Token and KeyToken, store before onContextEstablished (1500 changes)

My sum-up is:
Without "Cluster", #520 seems to me the best variant.
With "Cluster" it seems to me, that in the jungle of "possibilities", there is no conclusion about the exact definition of the behaviour of the `TokenProvider` (in context of the `MessageExchangeStore` an the `ObservationStore`) and therefore the "Cluster" seems to me very vague. That's my impression.

So, should we go for "sharpen" the "Cluster"? O r should we drop the "Cluster"?

**Comment by sbernard31** [comment]
The main reason about introducing ObservationStore was to support  failover and clustering.

At Leshan side, "cluster support" is the main reason why we start to use the 2.0.x version of californium.  
So I would go for sharpen the cluster, even if I'm not sure to see what you expect exactly.
Maybe you can precise which part of the cluster is still confused ?

Waiting I will try to read all those PRs :)

**Comment by boaks** [comment]
> Maybe you can precise which part of the cluster is still confused ?

FMPOV "releasing a Token" for "clustered observes" seems to be undefined, if the cancel is triggered from an other cluster node than the original requesting one. And this may be easy occur, if a RST is received from such an other node and will cause a "token leak" at the original requesting node.

**OK, I guess you will now just answer, that this is what you always wrote :-),** 

but I still feel, that the added complexity of variant #519 is not reasonable.

I would therefore prefer, to change the token generation:
Proposal:

For none observe requests:
1. Generate token by provide indicating local scope. 
    (the generator may implement this as random generator, set bit 0 to 1 (fix)).
2. Add the exchange with `putIfAbsent` to the local MessageExchangeStore.
3. if 2 fails, repeat from 1.

For observe requests:
1. Generate token by provide indicating global scope. 
    (the generator may implement this as random generator, set bit 0 to 0 (fix)).
2. Add the exchange/observation to the clustered ObservationStore also with 
     `putIfAbsent` semantic, (for redis I would go for the get / NX. The resulting small GAP should be acceptable, the probability of a clash of a 8 byte random token should be very low).
3. if 2 fails because the key exists, repeat from 1.
 
The token provider will only generate tokens, not maintain tokens.

**Comment by sbernard31** [comment]
> OK, I guess you will now just answer, that this is what you always wrote :-),

=> ❤️ 

> but I still feel, that the added complexity of variant #519 is not reasonable.

I don't see so much complexity, most of the change are because we add an EndpointContext parameter and we change the type of KeyToken and some optional modification (like Userinfo)
But maybe I missed some issue like the cancel one or the CoapObserveRelation one. (https://github.com/eclipse/californium/issues/500#issuecomment-353378579)

I still feel this is more in the CoAP spirit and this could allow us to remove the crappy `setContext` on `ObservationStore`.

> I would therefore prefer, to change the token generation.

Even if I rather prefer the other solution, this token generation proposal should work too.

**Comment by boaks** [comment]
#521 

Plain Token, with token generator and putIfAbsent.

The "complexity" in #519 is for me, that the changes will require more attention in the next time to ensure, that they doesn't introduce a new issue.

> I still feel this is more in the CoAP spirit and this could allow us to remove the crappy setContext on ObservationStore.

For the (dtls) server side it should be possible to have a empty implementation for that.
If we want to remove it completely, I would go for a simple "connect" (send a empty RawData, if the context is not "filled"). But this would be an other issue/PR.

**Comment by sbernard31** [comment]
> The "complexity" in #519 is for me, that the changes will require more attention in the next time to ensure, that they doesn't introduce a new issue.

I agree, but the real question is : "is it worthy ?"
But let's #519 aside and let's explore the #521 way.

>  I would go for a simple "connect" 

About adding 'connect', I don't know :/, I think 'connect' come with 'reconnect', 'abbreviated connect', 'close', 'not conntected error' ... I don't think this is a bad thing but I feel this is even more impacting for Californium than #519. This changes Californium paradigm a lot.
So, I think we can live with the "crappy" setContext at least at short/mid term. so let's removing "setContext" aside too.

Now let's talk about #521, I will comment the #521 PR to not flood this thread  so much.

**Comment by boaks** [comment]
The term "connect" is very much overloaded.

In my idea, it's only about having a "partially filled" endpoint context (related on the required information of an endpoint context matcher), and to fill it using a empty `RawData` to trigger the Connector to "fill in its information" (without sending application data) and calls back (as californium does it right now). No "close" and "no reconnect", but if you can't establish a DTLS session you will receive an error (but this will alos be the case in the current implementation) . 
But that would only be required on the dtls client side, if the endpoint context would be based on the DTLS session and cluster observe/notify would be required to be optimized. I don't see this in the near or middle future.

A LWM2M server should always have a valid endpoint context with receiving the registration request of a LWM2M device. So the leshan-server should not be affected by such a "connect" at all.

**Comment by boaks** [comment]
Closed with PR #521
